File: backend/src/agents/tools/list_sub_agents.py
from __future__ import annotations

from typing import Any

from src.schemas.sandbox import ToolExecutionResult


LIST_SUB_AGENTS_TOOL_SCHEMA = {
    "type": "function",
    "function": {
        "name": "list_sub_agents",
        "description": "Returns a list of all available sub-agents that have been created by the user and are currently enabled. Use this to discover which sub-agents can be called.",
        "parameters": {
            "type": "object",
            "properties": {}
        }
    }
}


async def execute_list_sub_agents(*, arguments: dict[str, Any], **kwargs) -> ToolExecutionResult:
    sub_agents = kwargs.get("sub_agents", [])
    available = []
    for sa in sub_agents:
        if sa.get("enabled", True):
            available.append({
                "name": sa.get("name", ""),
                "description": sa.get("description", ""),
            })

    return ToolExecutionResult(
        ok=True,
        data={
            "available_sub_agents": available,
        },
    )

File: curro-ai/src/agents/tools/list_sub_agents.ts
import { z } from "zod";
import { defineTool, type ToolContext, type ToolResult } from "./types.js";

const schema = z.object({});

export const listSubAgentsTool = defineTool({
  name: "list_sub_agents",
  description:
    "Returns a list of all available sub-agents (name and description). Use this to discover which " +
    "specialized sub-agents can be delegated work with call_sub_agent.",
  schema,
  label: () => "List Sub-Agents",
  async execute(_args, ctx: ToolContext): Promise<ToolResult> {
    const available = ctx.subAgents?.available() ?? [];
    return {
      ok: true,
      data: {
        available_sub_agents: available.map((a) => ({
          name: a.name,
          description: a.description,
        })),
      },
    };
  },
});
